staging/src/k8s.io/apiserver/pkg/apis/audit/types.go

```diff
@@ -39,6 +47,22 @@ const (
 	LevelRequestResponse Level = "RequestResponse"
 )
 
+// Stage defines the stages in request handling that audit events may be generated.
+type Stage string
+
+// Valid audit stages.
+const (
+	// The stage for events generated as soon as the audit handler receives the request, and before it
+	// is delegated down the handler chain.
+	StageRequestReceived = "RequestReceived"
+	// The stage for events generated once the response headers are sent, but before the response body
+	// is sent. This stage is only generated for long-running requests (e.g. watch).
+	StageResponseStarted = "ResponseStarted"
+	// The stage for events generated once the response body has been completed, and no more bytes
+	// will be sent.
+	StageResponseComplete = "ResponseComplete"
```

Comment:
What about the panic handler? Technically a panic can happen in any phase. We could add another stage for panic.